### DISCUSSION TITLE
Accessing iso file after booting (switching kernel) [memdisk, keep, keeppxe, sanboot]

### DISCUSSION THREAD

**Discussion by Ninos** [ipxe/ipxe#912] [General]
Hey my friends,

sorry to bother u with questions, but my knowledge about kernel-based things is a little limited...

Currently I'm using `sanboot` to boot iso-file over network. It's working great just with the restriction after switching to new kernel it's not possible to access iso-file anymore.

I understand, that switching kernel "replaces" everything before, so e.g. network stack get lost and all the other stuff done before. But I read something about the ´--keep´ option while booting via `sanboot`. In your documentation you're writing:
>  This drive will be automatically detached if booting then fails, unless the --keep option is specified.

What does that mean exactly? With option `--keep` enabled it's possible to access iso-file after switching to protected mode (int16 to int32 [whatever that exactly means :) ])? If yes, how is it possible?

Currently I'm using following ipxe commands:
```
#!ipxe

set keep-san 1 # Just to be sure, think this is the old way & does the same as --keep option?
sanboot --no-describe --drive 0x81 --keep http://example.tld/url/to/image.iso
```

My iso-file is configured with syslinux, the configuration file looks something like that:
```
LABEL lts
MENU LABEL Linux lts
KERNEL /boot/vmlinuz-lts
INITRD /boot/intel-ucode.img,/boot/amd-ucode.img,/boot/initramfs-lts
APPEND ... keeppxe
```
As I understood correctly in first step (`sanboot`) you're chainloading to iso-file (syslinux, grub, whatever defined), after that syslinux takes place and replaces kernel + initrd. With kernel option `keeppxe` syslinux should keep the pxe capabilities [see also [memdisk doc](https://wiki.syslinux.org/wiki/index.php?title=MEMDISK)]. So after that I tried finding iso-file using `memdiskfind` command from syslinux-package, but I cannot find any. I also cannot use something like that:
```
LABEL lts
MENU LABEL Linux lts
KERNEL memdisk
INITRD http://example.tld/url/to/image.iso
APPEND ... iso
```
Because otherwise I need two iso-files, first for loading syslinux with hardcoded url in `INITRD`, second for loading real system (and I'm not sure if this will work, because of may higher restrictions...)

**So my main question is:**
Is it possible (mainly in newer kernels with higher security, don't know if this matters..) to netboot from iso file and then access same iso inside protected kernel without adding contents to initrd-file or other shitty solutions? If yes, what am I doing wrong? :/ If no, is it technically possible (e.g. adding some functionality to ipxe), or restricted due to force majeure (security or whatever)?

Thx a lot!

**Comment by NiKiZe** [comment]
Generally you will not be able to boot ISO images for anything that switches to its own network stack.

sanboot is primarly made for iSCSI which uses the iBFT ACPI table (created by iPXE) to hand over the connection to the data, this standard is for iSCSI only, and also requires the booting system to be able to use iBFT.

With memdisk and "working" kernel, what happens is that whole file is loaded into memory, and special code in the kernel or userspace "searches all of the memory to find any memdisk data" when found it presents that as a block devices, and locks that memory (hopefully it was not corrupted by any process running before).

When PXE booting, don't use ISOs, extract the kernel and initrd, provide options for the kernel and initrd script to use any 2nd level filesystem.

Or load all the data from the start (if not to huge), such as explained and used at https://github.com/NiKiZe/Gentoo-iPXE

There is several other similar questions and explanations already, which might give other contexts, and might worth a read. (to many to even list them)

**Comment by Ninos** [comment]
Hmm I see, thank you very much for your fast help! I wanted to prevent extracting iso file before booting (some additional steps are then needed)...
Is it possible to define some variables before sanbooting iso file via http (or whatelse) and then use them in bootloader before switching to new kernel + initrd (e.g. in syslinux.cfg something like `APPEND ... isofile=${isofile_url}`)?

**Reply by NiKiZe** [reply]
When using iPXE consider ISOs and syslinux to be "the wrong direction"

You can define variables in iPXE but those are only available in iPXE.
If you add anything to the kernel cmdline, in linux that is available for parsing, and is commonly used to parse out 2nd stage images and downloaded, but each distro does this differently.

If you wouldn't mind sharing what you are trying to do, as in which distro you are trying to boot, or any other information, then maybe someone can better help with that particular challenge?
netboot.xyz have several examples.

Going back to the Gentoo example, I would like for them to distribute the required files available on their own to remove the need for extracting the ISO.
A different approach would be to use a webserver that can extract the ISO on the fly eg. https://github.com/ipxe/osbootd (No documentation and in Proof of concept stage, but should be working for those distros)

**Comment by Ninos** [comment]
> You can define variables in iPXE but those are only available in iPXE.

Ok thx, just to understand, the only possibility (in theory) to transfer data down (from pxe to syslinux) is to chainload them earlier on "disk"?

> If you wouldn't mind sharing what you are trying to do, as in which distro you are trying to boot, or any other information, then maybe someone can better help with that particular challenge?

In short:
I want to pre-build a bootable iso (the distro doesn't matter, currently alpine) and then just boot the iso via pxe in generic way (without defining distro-specific pxe-options [kernel, initrd, append, ...]).
The main problem is, that some distros needs some files while booting which are stored inside iso (e.g. apks, additional drivers, ...), which are get lost. Other things are already working fine just using sanboot, so I hoped also to find a possible/hookable way to access iso-content under protected kernel :-)

Thx for your link, exactly what I wanted to avoid: An additional step between :( But nice for inspirations! ❤️

**Comment by Ninos** [comment]
Solved it!

Instead of booting iso from ipxe, we need to load iso into initrd and later chain to memdisk kernel. For that just:
1. Add `memdisk` kernel to your webserver (file can be found e.g. under `/usr/share/syslinux/memdisk` in alpine distro after installing syslinux package).
2. Boot from ipxe with following commands:
   ```
   #!ipxe

   initrd http://example.tld/path/to/image.iso
   chain /memdisk iso
   ```

After chaining to memdisk you get a nice download progress bar and after success memdisk chains to iso. The **distribution needs to support memdisk**, for Alpine I already [created a patch](https://gitlab.alpinelinux.org/alpine/mkinitfs/-/merge_requests/115) :-)

PS: I've changed the title a bit, so may other will find their answers a bit faster...
PS: Don't try to use `sanboot`, just give up going the direct way, it will not work. ISO is not fully loaded into ram & there's no "bridge" from pxe to protected kernel, so after loosing network stack (switching kernel), you cannot access iso anymore. Also don't try to use `kernel` for chaining to memdisk, use `chain` instead. Parameters `keeppxe` & `vmalloc` are not needed (tested with an 800mb iso file & alpine).

**Reply by NiKiZe** [reply]
This does not work in EFI, consider it a hack rather than "a solution"

It would be much better if distros supported the kernel+initrd+squashfs approach, but that does require some memory (but should be less then the iso+memdisk)

**Reply by Ninos** [reply]
Ahhhhh I see....

Thx for your hint! Then I'll first work with legacy BIOS (memdisk) and later on implement some on-the-fly unpack ISO solution you preferred before (with usage of kernel & initrd instead of iso).

Sorry to ask some stupid questions, but want to know if life can gets easier here. As understood correctly sanboot (is sanboot working with efi/uefi btw?) just maps file requests to ranged http content header requests on ISO file, so ISO don't need to stay in whole RAM (RAM has some disadvantages, switching kernel can lead to corruption because of may overwrites).
Isn't it possible to create this mapping as new sys device, which then can be used by protected kernel? The "cache/already downloaded file in ram" (if something like that exists) then needs be cleared because of may memory corruption, also network stack is required on booted system & to be post-used by new created device, but I like the idea translating file requests to http range requests on-the-fly.
Is it technically (may after some kernel patches or custom kernel modules) possible or just impossible because nothing (only kernel options) can be passed technically to new kernel after replacing it?

I already wanted to donate to this great project (btw some informations about that are missing on your website :(), if technically possible I'm willing to pay for such implementation (if it's also in your interests) :-)

Otherwise I'll find my way, you already helped me a lot 👼

**Reply by NiKiZe** [reply]
sanboot works with EFI yes.

It would be nice to use initrd for part of iso, #430 is for that request but it is not trivial to get right.

You can already load any file into kernel space using initrd and cpio. What is needed is for the init scripts to detect and use that file. So corrected initrd scripts is what you need, again see the Gentoo example, I also implemented this for earlier versions of systemrescuecd which was merged.

If you have features you want implemented then mcb30 can do it for a fee, you will have to reach out on vendor support for that. Donations is not accepted. (But maybe you will be able to buy him a beer if you meet in person :D)

**Reply by Ninos** [reply]
> You can already load any file into kernel space using initrd and cpio. What is needed is for the init scripts to detect and use that file. So corrected initrd scripts is what you need, again see the Gentoo example, I also implemented this for earlier versions of systemrescuecd which was merged.

Yeah, in first step I tried creating kind of a fakerom (copying all required files into initrd), but performance is worse. You mean instead of copyiing files directy into initrd, it's better to create a squashfs and add this file into initrd-file, after that mounting it in custom init script of distro (may needs patch for alpine)?
As I understood correctly it's not working like faking a rom (mounted dir) by adding squashfs-file to iso directly and loading it with initrd-command? It will be "overlayed" nearly the same as other initrd-files, it's just faster in read performance & read-only? (Hope I understood your Gentoo-examples correctly :D)

**EDIT:**
Think I missunderstood squashfs a bit. After adding file to initrd command (e.g. something like this `INITRD /boot/intel-ucode.img,/boot/amd-ucode.img,/boot/initramfs-lts,/boot/filesystem.squashfs`) and booting iso, you get a file called initrd.image in root-dir, I think this is the squashfs-file. Currently I'm trying to mount it to a custom dir to be sure :D

**EDIT 2:**
Nvm, seems I was blind before, it's a cpio file. The squashfs file must e.g. be downloaded in init script as current solution.

> It would be nice to use initrd for part of iso, https://github.com/ipxe/ipxe/issues/430 is for that request but it is not trivial to get right.
> If you have features you want implemented then mcb30 can do it for a fee, you will have to reach out on vendor support for that. Donations is not accepted. (But maybe you will be able to buy him a beer if you meet in person :D)

I'll write an email to vendor support, after implementing kind of this he deserves more than a beer! :-)

**Comment by stappersg** [comment]
See also https://github.com/ipxe/ipxe/discussions/828#discussioncomment-4293755 where the _"Normally you would not boot ISOs"_ is probably _"Normally you would not netboot ISOs"_
